docs/tutorials/05_kv-scheduler.md

```diff
@@ -0,0 +1,303 @@
+
+1. to configure all the proto-modelled data provided by the northbound, usually more than one binary API call is required
+2. to configure parameters, a parent item is required to be present
+2. some configuration items themselves are dependent on other and cannot be configured earlier
+
+ It means that binary API calls have to be called in an exact order. The VPP agent uses a KV scheduler component to ensure it with system of caching and configuration dependency handling. Every plugin configuring something dependent on other plugin's configs in certain way can be added to the KV scheduler and profit from its advantages. 
+ 
+ As a first step, we will use the special [proto model][1]. The model defines two simple messages - an `Interface` and a `Route` requiring some interface. The model demonstrates simple dependency between configuration items (since we need an interface to configure the route).  
+ 
+ In order to register our hello world plugin to the scheduler and work with the new model, we need a two new components - a **descriptor** and an **adapter** for every proto-defined type.
+ 
+ #### 1. Adapters
+ 
+ Let's start with adapters. The purpose of the adapter is to define conversion methods between our proto-defined type and bare `proto.Message`. Since this code fulfils the definition of a boilerplate, we will generate it. The generator is called `descriptor-adapter` and can be found in the [inside the KVScheduler plugin][2]. Build the binary file from the go files inside, and use it to generate two adapters for the `Interface` and `Route` proto messages:
+ 
+```
+descriptor-adapter --descriptor-name Interface --value-type *model.Interface --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+descriptor-adapter --descriptor-name Route --value-type *model.Route --import "github.com/ligato/vpp-agent/examples/tutorials/05_kv-scheduler/model" --output-dir "descriptor"
+```
+
+It is a good practice to add those commands to the plugin file with the `//go:generate` directives. Adapters were automatically created in the `adapter` directory within plugin folder.
+
+#### 2. Descriptor without dependency
+
+Another step is to define descriptors. We start with the interface descriptor, since there is no dependency on other messages. Let's create the `descriptors.go` so the code is outside of `main.go`, and define following code:
+
+```go
+type IfDescriptor struct {
+	log logging.PluginLogger
+}
+
+func NewIfDescriptor(logger logging.PluginLogger) *IfDescriptor {
```

Comment:
Btw. [here in this example](https://github.com/ligato/vpp-agent/blob/master/examples/kvscheduler/mock_plugins/ifplugin/descriptor/interface.go#L64-L92) I have come up with a different scheme for the descriptor constructor, which avoids the confusing GetDescriptor() method and [registration is straightforward](https://github.com/ligato/vpp-agent/blob/master/examples/kvscheduler/mock_plugins/ifplugin/ifplugin.go#L76-L81). Maybe we could use this approach going forward, gradually updating existing descriptors and already recommending it in the docs - what do you thing?

Reply:
This looks nice - especially the descriptor registration - and if all descriptor parameters are directly returned as functions inside the NewDescriptor() (as it was in this example), the descriptor struct is not even needed. I changed the example a bit to show both ways and mentioned it in the tutorial.